### ISSUE TITLE
al_get_monitor_info returns wrong values on Retina displays

### ISSUE THREAD

**Issue by fatcerberus** [liballeg/allegro5#923]
On macOS, `al_get_monitor_info` returns the wrong values when called on a Retina display, specifically it seems like `x2` and `y2` are half of what they're supposed to be.  As miniSphere uses this function to determine window scaling and position, this results in game windows which are too small and off-center:

<img width="1680" alt="screen_shot_2018-08-04_at_11 17 24_am" src="https://user-images.githubusercontent.com/4097841/43905878-f8192718-9bbf-11e8-853e-0b0c41410177.png">

As Allegro is otherwise Retina-aware, `al_get_monitor_info` should return the native resolution rather than the one the system is faking.

**Comment by SiegeLord** [comment]
This may have been fixed by https://github.com/liballeg/allegro5/commit/b48269bb4d81be055744077efe46bfd6bd0cb7a1#diff-aafb8ca502839c6d514dca74271e83d0 as part of the DPI work.

**Comment by fatcerberus** [comment]
Thanks for the tip, I’ll check with my Mac tester(s) to see if building Allegro from master helps.

**Comment by dos1** [comment]
@fatcerberus So is it fixed? :)

**Comment by fatcerberus** [comment]
This is still not fixed.

**Comment by pedro-w** [comment]
I was able to try this on an old retina MBP yesterday (a 2013 model running 10.15). Using ex_monitorinfo from the current source, the returned values were what I'd expect for the 'pixel' resolution, i.e. 2000-something pixels across. (I think Apple docs call this the 'backing resolution'). @fatcerberus have you got multiple displays? What version of MacOS are you using?

**Comment by pedro-w** [comment]
Just an aside (and I may be looking in the wrong place), the `screen_new` code in sphere looks a bit strange
https://github.com/fatcerberus/sphere/blob/4736d31d0204886cf071cb19aad3690583eaab39/src/neosphere/screen.c#L99-L104
doing some maths on `x_scale` and `y_scale` but they are ints, on the other hand there *are* float `x_scale` and `y_scale` in `struct screen` but these are never set by `screen_new`

**Comment by fatcerberus** [comment]
The code you pointed out is to decide scaling for low-resolution games (e.g. retro-style 320x240) and is therefore always an integer.  I’m not sure why I didn’t truncate explicitly though, that is a bit confusing.

When the window is first created, Allegro centers it properly.  What actually breaks on a Retina display (i.e. pretty much all MacBooks now) is here:
https://github.com/fatcerberus/sphere/blob/4736d31d0204886cf071cb19aad3690583eaab39/src/neosphere/screen.c#L536-L540

**Comment by pedro-w** [comment]
OK.
Sorry to keep going over this but, when you run `ex_monitorinfo` do you see the wrong values printed? (I ask because it works fine with my testing - just need to be sure we're looking in the right place)

**Comment by fatcerberus** [comment]
I’ll check later but despite what the OP says, I don’t think it’s actually the monitor info that’s wrong - best I can tell, it’s `al_set_window_position()` specifically that’s broken.

**Comment by pedro-w** [comment]
Hmm, that code doesn't seem to take any account of DPI
https://github.com/liballeg/allegro5/blob/4aa54e6c994af21bc63d8b593673ab3df62390f8/src/macosx/osxgl.m#L2200-L2217
@SiegeLord  should all those coords be in pixels?

**Comment by pedro-w** [comment]
Compare the resize code where it does seem use functions to convert from the Backing rectangle size
https://github.com/liballeg/allegro5/blob/4aa54e6c994af21bc63d8b593673ab3df62390f8/src/macosx/osxgl.m#L2074-L2090
If this is right, I can do a PR for `set_window_position`

**Comment by fatcerberus** [comment]
Wait, it doesn’t just use the backing resolution automatically for window position? That’s a bit odd.

**Comment by pedro-w** [comment]
I've definitely got something wrong - if my theory was right then ex_windows shouldn't work correctly on retina displays. It places two windows at 1/3 and 2/3 of the width, and if the screen size was the backing resolution but the position was the logical resolution then placing would be wrong.

**Comment by fatcerberus** [comment]
Okay so my current test machine is a 13" M1 MacBook Pro which, according to Apple's own spec sheet, has a 2560x1600 panel, but I see this in the debugger:
![image](https://user-images.githubusercontent.com/4097841/118294256-87284200-b4a8-11eb-8784-3becc6a63183.png)

I have no idea what to make of this.

**Comment by pedro-w** [comment]
I've been looking into this. Possibly you have your display set to 'scaled'? See for example the comment by robvas [here](https://www.reddit.com/r/mac/comments/991yj4/2018_15_mbp_is_not_2880x1800_its_3360x2100/)
Q. I've taken a full screenshot of the desktop and opened the image in PS and showed the actual resolution of it is 3360x2100 wtf is happened here, Apple announced that the resolution is 2880x1800
A. If you select the 1680 x 1050 or 1920 x 1200 scaling modes, Apple actually renders the desktop at 2x the selected resolution (3360 x 2100 or 3840 x 2400, respectively), scales up the text and UI elements accordingly so they aren’t super tiny (backing scale factor = 2.0), and downscales the final image to fit on the 2880 x 1800 panel.

In the image at the top, the window is exactly 640x480 pixels (as specified in the json file I assume) but possibly when it was scaled down on the screen it became smaller. I think. My head is spinning!

**Comment by fatcerberus** [comment]
My resolution is actually set to "Default for Display", which according to System Information "looks like" 1440x900 so that explains the 2880x1800 backing res.  Retina scaling is weird.  I think it's that "scaling back down" step that is causing all these issues, that doesn't normally happen even in DPI-aware situations (e.g. Windows never does this).

**Comment by pedro-w** [comment]
I think this means that 'pixel perfect' scaling is going to be difficult if not impossible for windowed mode - as you say it's the scaling back down. This is the last stage in the process and if that's not a integer there's not much we can do about it. The backing rect is [not a global thing](https://developer.apple.com/library/archive/documentation/GraphicsAnimation/Conceptual/HighResolutionOSX/Explained/Explained.html#//apple_ref/doc/uid/TP40012302-CH4-SW2), so maybe a fullscreen window would allow a backing which is an integer multiple of the screen's native resolution. There are some interesting comments/links from SDL-world here:
https://github.com/NagyD/SDLPoP/issues/173

**Comment by pedro-w** [comment]
Also the highdpi work didn't change things everywhere it needed to, I think. For example
https://github.com/liballeg/allegro5/blob/9de18401d51144071ee2943afdf7640e5d8e0287/src/macosx/osxgl.m#L2200-L2217
I'm looking into this

**Comment by fatcerberus** [comment]
I don’t care if my scale factor is ultimately not an integer - I just want to be able to center the window, which currently doesn’t work.

**Comment by pedro-w** [comment]
Well, in that case we're on the right track. I mentioned `al_set_window_position` [previously](https://github.com/liballeg/allegro5/issues/923#issuecomment-840584833) - the other key fact is that the window creation code doesn't call `set_window_position`, it has its own little block
https://github.com/liballeg/allegro5/blob/4aa54e6c994af21bc63d8b593673ab3df62390f8/src/macosx/osxgl.m#L1614-L1624
and this _does_ take the screen scaling factor into account.
I've got a half done patch but I need to check there aren't some other places where it's wrong.

**Comment by pedro-w** [comment]
Apologies for the delay, I will get back to this asap.

**Comment by pedro-w** [comment]
I did commit some code and then forgot to mention it. Here it is, I'm really unsure if it's right as I haven't the hardware to test. If someone could check it out that would be great.
https://github.com/pedro-w/allegro5/tree/issue-923

**Comment by SiegeLord** [comment]
I managed to get a second monitor. Turns out there's a number of bugs present... in particular, the Y axis wasn't being flipped correctly. More concerningly, there is a fundamental issue where the monitors only don't overlap in the overall XY plane if we use the scaled coordinates. If we use pixel coordinates, they end up overlapping. This isn't something OSX can help us with, we'll need to manually move the monitors around... I haven't tried this yet.
